sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -0,0 +1,214 @@
+ * types (i.e., byte, short, int and long). The rule checks to see if the literal `value` is
+ * within range `(min, max)`, where `min` and `max` are the minimum and maximum value of
+ * `fromType`, respectively. If this is true then it means we can safely cast `value` to `fromType`
+ * and thus able to move the cast to the literal side.
+ *
+ * If the `value` is not within range `(min, max)`, the rule breaks the scenario into different
+ * cases and try to replace each with simpler constructs.
+ *
+ * if `value > max`, the cases are of following:
+ *  - `cast(exp, ty) > value` ==> if(isnull(exp), null, false)
+ *  - `cast(exp, ty) >= value` ==> if(isnull(exp), null, false)
+ *  - `cast(exp, ty) === value` ==> if(isnull(exp), null, false)
+ *  - `cast(exp, ty) <=> value` ==> false
+ *  - `cast(exp, ty) <= value` ==> if(isnull(exp), null, true)
+ *  - `cast(exp, ty) < value` ==> if(isnull(exp), null, true)
+ *
+ * if `value == max`, the cases are of following:
+ *  - `cast(exp, ty) > value` ==> if(isnull(exp), null, false)
+ *  - `cast(exp, ty) >= value` ==> exp == max
+ *  - `cast(exp, ty) === value` ==> exp == max
+ *  - `cast(exp, ty) <=> value` ==> exp == max
+ *  - `cast(exp, ty) <= value` ==> if(isnull(exp), null, true)
+ *  - `cast(exp, ty) < value` ==> exp =!= max
+ *
+ * Similarly for the cases when `value == min` and `value < min`.
+ *
+ * Further, the above `if(isnull(exp), null, false)` is represented using conjunction
+ * `and(isnull(exp), null)`, to enable further optimization and filter pushdown to data sources.
+ * Similarly, `if(isnull(exp), null, true)` is represented with `or(isnotnull(exp), null)`.
+ */
+object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case l: LogicalPlan =>
+      l transformExpressionsUp {
+        case e @ BinaryComparison(_, _) => unwrapCast(e)
+      }
+  }
+
+  private def unwrapCast(exp: Expression): Expression = exp match {
+    case BinaryComparison(Literal(_, _), Cast(_, _, _)) =>
+      // Not a canonical form. In this case we first canonicalize the expression by swapping the
+      // literal and cast side, then process the result and swap the literal and cast again to
+      // restore the original order.
+      def swap(e: Expression): Expression = e match {
+        case GreaterThan(left, right) => LessThan(right, left)
+        case GreaterThanOrEqual(left, right) => LessThanOrEqual(right, left)
+        case EqualTo(left, right) => EqualTo(right, left)
+        case EqualNullSafe(left, right) => EqualNullSafe(right, left)
+        case LessThanOrEqual(left, right) => GreaterThanOrEqual(right, left)
+        case LessThan(left, right) => GreaterThan(right, left)
+        case _ => e
+      }
+
+      swap(unwrapCast(swap(exp)))
+
+    case BinaryComparison(Cast(fromExp, _, _), Literal(value, toType: IntegralType))
```

Comment:
Is it possible that in `Cast(fromExp, toType1, _)` and `Literal(value, toType2: IntegralType)`, `toType1` and `toType2` are different? Should we check? They will be eventually the same by other rules, but we want this rule kick in after they are the same type.

Reply:
Maybe just do 
```
BinaryComparison(Cast(fromExp, _: IntegralType, _), Literal(value, toType: IntegralType))
```
to make sure, and it will be readable as well.

Reply:
Good point. We are relying on that assumption but it is better to add a check.

Reply:
Yeah, without checking, we might break the query plan if the plan we get here is before type coercion.